src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2776,7 +2776,13 @@ ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
 					mono_jit_info_table_find_internal ((gpointer)data->stack_contents.stack_frames [frame_count], TRUE, FALSE);
 			}
 			mono_thread_info_set_tid (&adapter, ep_rt_uint64_t_to_thread_id_t (data->thread_id));
-			ep_write_sample_profile_event (sampling_thread, sampling_event, &adapter, &data->stack_contents, (uint8_t *)&data->payload_data, sizeof (data->payload_data));
+			uint32_t payload_data =
+#if BIGENDIAN
+				__builtin_bswap32 (data->payload_data);
+#else
+				data->payload_data;
+#endif
+			ep_write_sample_profile_event (sampling_thread, sampling_event, &adapter, &data->stack_contents, (uint8_t *)&payload_data, sizeof (payload_data));
```

Comment:
Rather than do these payload conversions here, it may be simpler to identify a big-endian arch at the reader end and handle it there. Doing it here will require you to find _every_ event, including user defined ones, with a payload and convert the contents to little-endian before it gets cast to a byte array. You'll either only have a few events with little-endian payloads, or you'll spend a lot more time converting everything and re-engineering all the code paths to swap bytes around.

Reply:
I think this is a question of scope. If @stefan-sf-ibm knows he won't care about EventSource generated events then the current approach handles it. @stefan-sf-ibm if you do care about those events then I could imagine a few ways to deal with it:
a. Do similar endian swapping loging in EventSource serializer and consider all events manually written via WriteEventCore as out of scope
b. Write some code that will parse event metadata to understand the schema of the payload and then endian swap all the fields (this handles the WriteEventCore case not handled in (a)
c. Make a breaking change to the file format allowing for big endian encodings and update format readers to accomodate it.

(b) adds more perf overhead, (c) probably creates a situation where tools based on sufficiently new TraceEvent support big endian data but older tools or non-TraceEvent based tools don't (I think we're up to at least three 3rd party implementations).

My inclination is to favor runtime complexity over file format complexity (favor the current PR approach or a/b options over the c option) assuming that big endian devices remain a tiny fraction of overall usage.

Reply:
I think that the current PR approach and the scope that it covers, payloads generated through https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/ClrEtwAll.man and emitted by Mono runtime, is a good start, I guess it mainly covers the initial TARGET_S390X needs, AFIK, that is the only big endian architecture currently running Mono out of dotnet/runtime (don't think CoreCLR supports any big endian platforms). If more support for custom payloads are needed in the future for big endian platforms, I believe we could revisit this, but current PR will at least make sure all Mono runtime emitted events are readable by existing tools.

Reply:
Please see my comment regarding the overhead for little-endian architectures in the above [comment](https://github.com/dotnet/runtime/pull/68648#discussion_r868882196).

Ideally we would have for nettrace format something similar like in the Common Trace Format used by LTTng where one can define globally or even per integer (probably not important for us) the [byte order](https://diamon.org/ctf/#spec4.1.3). This would have the least overhead while taking a trace which is performance critical. My experience with EventPipe is limited to the last couple of weeks where I basically only got the implementation running. So I'm not sure how important all this is and how it manifests in practice like skipping events due to high load only because of byte swaps and its side-effects from the implementation like non-inlined function calls and so on. I'm also not sure how likely it is to change the current format and in which time frame something like that could happen. If there is anything I can do I would be happy to jump in. Anyhow, in order to gain some experience this PR might suite as a starter where all Mono events are handled. Though, in the long run I'm still kind of worried about user events.

@josalem the TraceEvent library is also on my todo list. A nettrace file created with and without this PR is not convertible on a big-endian machine. With this PR it is at least convertible on a little-endian machine. However, I just gave the sample provider a chance yesterday and traced an example application and the trace looks broken to me, i.e., the flamegraph contains no function names. It looks like as I have missed something which I need to investigate ... setting this PR to WIP.

Reply:
Make sure you get the rundown events into the .nettrace files and that they are correct since they are needed for the  symbolication of callstacks it tools, they have the mapping between method id and the function names/signatures. You can look at all events included in the nettrace file in perfview, and on some, it will also display the payload, but not for all.